src/viewer/withFileUrl.jsx

```diff
@@ -0,0 +1,70 @@
+import React, { Component } from 'react'
+import { getDownloadLink } from 'cozy-client'
+import Spinner from 'cozy-ui/react/Spinner'
+
+import NoNetwork from './NoNetwork'
+
+const TTL = 6000
+
+const LOADING = 'LOADING'
+const LOADED = 'LOADED'
+const FAILED = 'FAILED'
+
+const withFileUrl = WrappedComponent => {
+  return class Wrapper extends Component {
+    state = {
+      status: LOADING,
+      downloadUrl: null
+    }
+
+    componentWillMount() {
+      this.loadDownloadUrl()
+    }
+
+    componentWillReceiveProps(nextProps) {
+      if (nextProps.file.id !== this.props.file.id) {
+        this.reset()
+      }
+    }
+
+    componentDidUpdate() {
+      if (this.state.status === LOADING && !this.timeout) {
+        this.loadDownloadUrl()
+      }
+    }
+
+    loadDownloadUrl() {
+      this.timeout = setTimeout(() => this.setState({ status: FAILED }), TTL)
+      getDownloadLink(this.props.file)
+        .then(url => {
+          this.clearTimeout()
+          this.setState(state => ({ downloadUrl: url, status: LOADED }))
+        })
+        .catch(() => {
+          this.clearTimeout()
+          this.setState({ status: FAILED })
+        })
+    }
+
+    clearTimeout() {
+      clearTimeout(this.timeout)
+      this.timeout = null
```

Comment:
Since you're resetting `this.timeout` to `null` here I think it would be better to also initialize it at the begining.

Reply:
Par défaut si tu crées un prototype, un objet ou une classe, tu peux appeler la propriété que tu veux qui vaudra `null` si elle n'est pas initialisée.
Est-ce que ça apporte vraiment de faire `this.timeout = null` dans un code d'initialisation comme le constructeur ?

Reply:
Je suis d'accord avec @enguerran d'autant plus qu'initialiser ces propriétés dans le constructeur nous fait perdre la légèreté de la def. du state initial en class property.

Reply:
> Est-ce que ça apporte vraiment de faire this.timeout = null dans un code d'initialisation comme le constructeur ?

A mon sens oui — pas pour le fonctionnement, je sais bien que tu peux appeler une propriété non déclarée sans problème. Mais je trouve ca très utile pour la lecture par les humains, parce que tu a tout de suite la visibilité sur les props qui seront utilisées plus tard, au lieu de les découvrir une par une. C'est à mon avis la même raison qui fait qu'on initialise tout le state.
> […] initialiser ces propriétés dans le constructeur nous fait perdre la légèreté de la def. du state initial en class property.

Non, tu peux les déclarer de la même façon que le state:

```
class Lol extends Component {
  state = { … }
  maProp = null
  timeout = null

  premiereMéthodeLol () { … }
}
```

Reply:
Initialiser le state c'est comme initialiser une machine à états finis (cf. https://css-tricks.com/robust-react-user-interfaces-with-finite-state-machines/)

Ici, le smell, c'est qu'on a des propriétés qui accompagnent le composant (`timeout` et `img`) mais ces propriétés n'ont rien à voir avec le state du composant. Ils accompagnent le composant et je pense qu'on doit pouvoir les extraire du composant pour n'avoir qu'un composant étant une machine à état (et donc découplé, et donc facilement testable, etc.).

Mais chaque chose en son temps, ce refacto se fera quand on aura chopé un pattern qui va bien.